Expert advice ssl v tls (#1088)
FusionAuth/fusionauth-site: site/learn/expert-advice/security/ssl-vs-tls.md

---
layout: advice
title: SSL vs. TLS
description: In this blog, you’ll learn the difference between SSL and TLS, when they are used, and why you should use them.
author: Akira Brand
image: advice/steps-secure-auth-system/expert-advice-ten-steps-to-secure-your-authentication-system-header-image.png
category: Security
date: 2021-11-12
dateModified: 2021-11-12
---

In this blog, you’ll learn the difference between SSL and TLS, when they are used, and why you should use them.

*What are SSL and TLS?*

SSL (secure socket layer) and TLS (transport layer security) are a set of protocols that are used when a computer is communicating with an external server to get information.  The protocols ensure secure communication over the web by encrypting and securing any information sent over the world wide web.   

*So what’s the difference between the two?*

> There isn’t a difference, actually.  TLS actually replaced SSL, but since SSL has been around for longer, people still refer to TLS as SSL. Confused yet?  Read on!

Reviewer: Please use one space between sentences.

*When do you use SSL? I mean TLS, I mean..wait, what?*

> You use TLS, formally known as SSL, mainly when you send data between a client and a server.

Reviewer: ```suggestion
You use TLS, formerly known as SSL, mainly when you send data between a client and a server.
```

*Ok, I get it. TLS used to be SSL. But how does it work?*

SSL and TLS work by establishing a handshake protocol between a client and a server.  Once that handshake protocol is established, the client and server decide on a secret key that will be used to encrypt and decrypt all communications between them.

In this way, an eavesdropper cannot read or manipulate any of the data sent between the client and the server, they can only see the endpoints.

Now, the overall process of this handshake protocol looks like this:

> 

Reviewer: ```suggestion
```

1. The client sends a “clienthello” message to the server.  This message includes information like SSL/TLS version, and the cryptographic algorithm and data compression methods that the client can support.  
2. The server responds with a “serverhello” message. This message includes cryptographic algorithms that the server has chosen from the list provided by the client. In this way, an algorithm that both sides support is chosen. The server also sends a session Id, its digital certificate, and its public key.

> 3. The client contacts the server’s certificate authority to verify the validity of the server’s certificate. This confirms the authenticity of the web server.

Reviewer: "The client contacts the server’s certificate authority"

Is this true? I thought that it would look at the list of certificates it knows about and see if the certificate was signed by one of them.

 http://certificate.fyicenter.com/925_Trusted_Root_Certification_Authorities_in_Google_Chrome.html for example.

4. A secret key is exchanged.  The client sends a shared secret key to be used between the client and the server.  The secret key is encrypted with the server’s public key.
5. The client sends a “finished” message to the server, which is encrypted with the secret key.  This confirms that the client section of the handshake protocol is completed.
6. The server responds with a “finished” message, encrypted with the secret key. This indicates the server part of the handshake protocol is finished.  
7. The client and server now exchange messages that are symmetrically encrypted with the shared secret key.

{% include _image.liquid src="/site/_diagrams/learn/expert-advice/ssl-vs-tls/ssl-handshake.plantuml" alt="SSL/TLS handshake protocol" class="img-fluid" figure=false %}  

> Tl;dr, the entire purpose of the handshake is to verify the validity of the server, as well as securely exchange a shared secret key, which is used to encrypt and decrypt messages sent between the client and the server.

Reviewer: ```suggestion
Tl;dr, the entire purpose of the handshake is to verify the validity of the server, as well as securely exchange a shared secret key, which is used to encrypt and decrypt messages sent between the client and the server. Encryption with a shared secret is much faster, but the handshake with the asymmetric keys is required to safely share that secret.
```

*Why should I care?*

> Imagine you are sending a very important message to someone through the mail, that if it was intercepted and read along the way, would have disastrous consequences for your life.  You would need some way of coding the message so if someone finds it, they can’t understand it (encryption) and putting it in an envelope so that the message is secure (securing the communication).  In the same way, information sent over the internet is assumed to need encryption and security, because all kinds of data needs to be sent over the web: credit card numbers, secret love notes, health information, you name it. It all needs to be encrypted and secured, and the way that happens currently is with SSL/TLS.

Reviewer: ```suggestion
Imagine you are sending a very important message to someone through the mail, that if it was intercepted and read along the way, would have disastrous consequences for your life.  You would need some way of coding the message so if someone finds it, they can’t understand it (encryption) and putting it in an envelope so that the message is secure (securing the communication).  In the same way, information sent over the internet needs encryption and security, because all kinds of data is sent over the web: credit card numbers, secret love notes, health information, you name it. It all needs to be encrypted and secured, and the way that happens currently is typically SSL/TLS.
```

3. The client contacts the server’s certificate authority to verify the validity of the server’s certificate. This confirms the authenticity of the web server.
4. A secret key is exchanged.  The client sends a shared secret key to be used between the client and the server.  The secret key is encrypted with the server’s public key.
5. The client sends a “finished” message to the server, which is encrypted with the secret key.  This confirms that the client section of the handshake protocol is completed.
6. The server responds with a “finished” message, encrypted with the secret key. This indicates the server part of the handshake protocol is finished.  
7. The client and server now exchange messages that are symmetrically encrypted with the shared secret key.

{% plantuml source: _diagrams/learn/expert-advice/ssl-vs-tls/ssl-handshake.plantuml, alt: "The SSL/TLS handshake protocol." %}

Tl;dr, the entire purpose of the handshake is to verify the validity of the server, as well as securely exchange a shared secret key, which is used to encrypt and decrypt messages sent between the client and the server.

*Why should I care?*

> Imagine you are sending a very important message to someone through the mail, that if it was intercepted and read along the way, would have disastrous consequences for your life.  You would need some way of coding the message so if someone finds it, they can’t understand it (encryption) and putting it in an envelope so that the message is secure (securing the communication).  In the same way, information sent over the internet is assumed to need encryption and security, because all kinds of data needs to be sent over the web: credit card numbers, secret love notes, health information, you name it. It all needs to be encrypted and secured, and the way that happens currently is with SSL/TLS.

Reviewer 2: > Imagine you are sending a very important message to someone through the mail, that if it was intercepted and read along the way, would have disastrous consequences for your life.

Is this a run-on sentence? (I'm not a grammar expert)

Suggestion:

> Imagine you are sending a very important message to someone through the mail. If this email was intercepted or read along the way, it would have disastrous consequences for your life. 

Or 

> Imagine you are sending a very important message to someone through the mail, that if  intercepted or read along the way, would have disastrous consequences for your life.

---
layout: advice
title: SSL vs. TLS

> description: In this blog, you’ll learn the difference between SSL and TLS, how they are used, and why you should use them.

Reviewer: Not a blog. Maybe "post" or "article"

author: Akira Brand
image: advice/steps-secure-auth-system/expert-advice-ten-steps-to-secure-your-authentication-system-header-image.png
category: Security
date: 2021-11-12
dateModified: 2021-11-12
---

In this blog, you’ll learn the difference between SSL and TLS, when they are used, and why you should use them.

*What are SSL and TLS?*

> SSL (secure socket layer) and TLS (transport layer security) are a set of protocols that are used when a computer is communicating with an external server to get information. The protocols ensure secure communication over the web by encrypting and securing any information sent over the world wide web.   

Reviewer: ```suggestion
SSL (secure socket layer) and TLS (transport layer security) are a set of protocols that are used when a computer is communicating with an external server to get information. The protocols ensure secure communication over the web by encrypting and securing any information sent with the HTTP protocol.
```

*So what’s the difference between the two?*

> There isn’t a difference, actually. TLS replaced SSL, but since SSL has been around for longer, people still refer to TLS as SSL. Confused yet? Read on!

Reviewer: I also am confused. You are saying they are the same, but one replaced the other? You might want to say "there isn't a functional difference" or something similar.

*When do you use SSL? I mean TLS, I mean..wait, what?*

You use TLS, formally known as SSL, mainly when you send data between a client and a server.

*Ok, I get it. TLS used to be SSL. But how does it work?*

SSL and TLS work by establishing a handshake protocol between a client and a server. Once that handshake protocol is established, the client and server decide on a secret key that will be used to encrypt and decrypt all communications between them.

In this way, an eavesdropper cannot read or manipulate any of the data sent between the client and the server, they can only see the endpoints.

Now, the overall process of this handshake protocol looks like this:

1. The client sends a “clienthello” message to the server. This message includes information like SSL/TLS version, and the cryptographic algorithm and data compression methods that the client can support.  
2. The server responds with a “serverhello” message. This message includes cryptographic algorithms that the server has chosen from the list provided by the client. In this way, an algorithm that both sides support is chosen. The server also sends a session Id, its digital certificate, and its public key.

> 3. The client verifies the validity of the server’s certificate with the entity that issued it. It does so by first checking to ensure that the server's certificate is not expired and the domain name or IP address on the certificate matches the server's information. Then, the client attempts to verify that the server's certificate has been signed by the certificate authority who authorized it. It does so by verifying that the signature on the certificate has been signed by the certificate authority's public key.

Reviewer: What is a certificate authority?

4. A secret key is exchanged. The client sends a shared secret key to be used between the client and the server. The secret key is encrypted with the server’s public key.
5. If required, the server verifies the client's certificate.
6. The client sends a “finished” message to the server, which is encrypted with the secret key. This confirms that the client section of the handshake protocol is completed.
7. The server responds with a “finished” message, encrypted with the secret key. This indicates the server part of the handshake protocol is finished.  
8. The client and server now exchange messages that are symmetrically encrypted with the shared secret key.

{% plantuml source: _diagrams/learn/expert-advice/ssl-vs-tls/ssl-handshake.plantuml, alt: "The SSL/TLS handshake protocol." %}

Tl;dr, the entire purpose of the handshake is to verify the validity of the server. The handshake also securely exchanges a shared secret key, which is used to encrypt and decrypt messages sent between the client and the server.

*Why should I care?*

> Imagine you are sending a very important message to someone through the mail, that if it was intercepted and read along the way, would have disastrous consequences for your life.  You would need some way of coding the message so if someone finds it, they can’t understand it (encryption) and putting it in an envelope so that the message is secure (securing the communication).  In the same way, information sent over the internet is assumed to need encryption and security, because all kinds of data needs to be sent over the web: credit card numbers, secret love notes, health information, you name it. It all needs to be encrypted and secured, and the way that happens currently is with SSL/TLS.

Reviewer: Does it make sense to have this be at the start? "why should you care" is usually a great way to start a post.
